packages/airnode-node/src/coordinator/calls/chain-limits.ts

```diff
@@ -0,0 +1,61 @@
+import flatMap from 'lodash/flatMap';
+import groupBy from 'lodash/groupBy';
+import map from 'lodash/map';
+import zip from 'lodash/zip';
+import * as logger from '../../logger';
+import { Config, ProviderStates, PendingLog, GroupedRequests } from '../../types';
+import { flattenRequests, groupRequests } from '../../requests/grouping';
+import { sortRequests } from '../../requests/sorting';
+
+function flattenAndSortRequests(grouped: GroupedRequests) {
+  return sortRequests(flattenRequests(grouped));
+}
+
+export function applyChainLimits(config: Config, providerStates: ProviderStates) {
+  const logs: PendingLog[] = [];
+
+  const providerRequests = providerStates.evm
+    .map((evmState, index) => {
+      return { index, chainId: evmState.settings.chainId, requests: flattenAndSortRequests(evmState.requests) };
+    })
+    // Sort by number of request ascendingly
+    .sort((p1, p2) => p1.requests.length - p2.requests.length);
+  const groupedProviderRequests = groupBy(providerRequests, 'chainId');
+  const chainRequests = map(groupedProviderRequests, (group, chainId) => ({
+    chainId,
+    chainLimit: config.chains.find((chain) => chain.id === chainId)!.maxConcurrentTransactions,
+    requestsGroups: group,
+  }));
+
+  // Apply the chain request limit by dropping requests which exceed the limit. We do this mutably to increase
+  // coordinator performance
+  chainRequests.forEach((chain) => {
+    const groupsCount = chain.requestsGroups.length;
+    let remainingChainLimit = chain.chainLimit;
+
+    chain.requestsGroups.forEach((group, groupIndex) => {
+      const requestCount = group.requests.length;
+      const unprocessedGroupsCount = groupsCount - groupIndex;
+
+      const allowedRequestCount = Math.min(Math.floor(remainingChainLimit / unprocessedGroupsCount), requestCount);
```

Comment:
Would you be able to add a comment here explaining this? It's not so obvious to me and it's a core part of how this works